app/flatpak-builtins-search.c

```diff
@@ -0,0 +1,196 @@
+      GFile *appstream_file = g_ptr_array_index (appstream_dirs, i);
+      g_autoptr(GError) error = NULL;
+      as_store_from_file (store, appstream_file, NULL, cancellable, &error);
+      if (error)
+        {
+          // We want to ignore this error as its harmless and valid
+          // NOTE: appstream-glib doesn't have granular file-not-found error
+          if (!g_str_has_suffix (error->message, "No such file or directory"))
+            g_warning ("%s", error->message);
+        }
+    }
+
+  return store;
+}
+
+typedef struct MatchResult {
+  AsApp *app;
+  guint score;
+} MatchResult;
+
+static void
+match_result_free (MatchResult *result)
+{
+  g_object_unref (result->app);
+  g_free (result);
+}
+
+static MatchResult *
+match_result_new (AsApp *app, guint score)
+{
+  MatchResult *result = g_new (MatchResult, 1);
+  result->app = g_object_ref (app);
+  result->score = score;
+  return result;
+}
+
+static int
+compare_by_score (MatchResult *a, MatchResult *b, gpointer user_data)
+{
+  // Reverse order, higher score comes first
+  return (int)b->score - (int)a->score;
+}
+
+static void
+print_app (MatchResult *res, gpointer user_data)
+{
+  /* TODO:
+   * - Make formatting pretty
+   * - AsApp doesn't expose the flatpak repository name
+   * - Get localized summary
+   */
+  g_print("%-45s | %s\n", as_app_get_id_filename (res->app), as_app_get_comment (res->app, "C"));
+}
+
+gboolean
+flatpak_builtin_search (int argc, char **argv, GCancellable *cancellable, GError **error)
+{
+  g_autoptr(GOptionContext) context = g_option_context_new (_("TEXT - Search for text"));
+  g_option_context_set_translation_domain (context, GETTEXT_PACKAGE);
+
+  if (!flatpak_option_context_parse (context, options, &argc, &argv, FLATPAK_BUILTIN_FLAG_NO_DIR,
+                                     NULL, cancellable, error))
+    return FALSE;
+
+  if (argc < 2)
+    return usage_error (context, _("TEXT must be specified"), error);
+
+  const char *search_text = argv[1];
+  g_autoptr(AsStore) store = store_new (cancellable);
+  g_autoptr(GPtrArray) apps = as_store_get_apps (store);
+  GSList *matches = NULL;
+
+  for (guint i = 0; i < apps->len; ++i)
+    {
+      AsApp *app = g_ptr_array_index (apps, i);
+      const guint score = as_app_search_matches (app, search_text);
+      if (score != 0)
+        {
+           MatchResult *result = match_result_new (app, score);
+           matches = g_slist_insert_sorted_with_data (matches, result,
+                       (GCompareDataFunc)compare_by_score, NULL);
+        }
+    }
+
+  if (matches != NULL)
+    {
+      g_print ("%-45s | %s\n", _("Application ID"), _("Description"));
+      for (guint i = 0; i < glnx_console_columns (); ++i)
+        g_print ("-"); // Probably a smarter way
+      g_print ("\n");
+      g_slist_foreach (matches, (GFunc)print_app, NULL);
+      g_slist_free_full (matches, (GDestroyNotify)match_result_free);
+    }
+  else
+    {
+      g_print ("%s\n", _("No matches found"));
```

Comment:
I wonder if we should have a non-zero exit code when there aren't any matches, like `grep` does.

Reply:
In the short term I want to consider the output not scripting friendly as it may potentially change, so at that point the exit code isn't useful yet either.